### PR TITLE
Topic/oshmem mem prefetch v2.x

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe @miked-mellanox @jladd-mlnx 
please take a look

### CODE REVIEW

File: oshmem/mca/spml/ucx/spml_ucx.c [hunk 1]
```diff
@@ -288,7 +288,7 @@ int mca_spml_ucx_add_procs(ompi_proc_t **procs, size_t nprocs)
                             &ep_params,
                             &mca_spml_ucx.ucp_peers[i].ucp_conn);
         if (UCS_OK != err) {
-            SPML_ERROR("ucp_ep_create failed!!!\n");
+            SPML_ERROR("ucp_ep_create failed: %s\n", ucs_status_string(err));
```
Comment (yosefe) [inline oshmem/mca/spml/ucx/spml_ucx.c]: we can remote the \n

File: oshmem/mca/spml/ucx/spml_ucx.c [hunk 2]
```diff
@@ -539,7 +567,7 @@ int mca_spml_ucx_fence(void)
 
     err = ucp_worker_flush(mca_spml_ucx.ucp_worker);
     if (UCS_OK != err) {
-        SPML_ERROR("fence failed");
+        SPML_ERROR("fence failed: %s", ucs_status_string(err));
```
Comment (yosefe) [inline oshmem/mca/spml/ucx/spml_ucx.c]: indentation

File: oshmem/mca/spml/ucx/spml_ucx.c [hunk 3]
```diff
@@ -552,7 +580,7 @@ int mca_spml_ucx_quiet(void)
...
```
Comment (yosefe) [inline oshmem/mca/spml/ucx/spml_ucx.c]: indentation

File: oshmem/mca/memheap/buddy/memheap_buddy.c [hunk 4]
```diff
@@ -469,7 +469,7 @@ static int _do_alloc(uint32_t order,
     }
 
     *p_buff = (void*) addr;
-    /* no barrier because it is not required by spec! */
+    MCA_SPML_CALL(memuse_hook(addr, 1<<order));
```
Comment (miked-mellanox) [inline oshmem/mca/memheap/buddy/memheap_buddy.c]: 1LL<<order
2nd param is size_t or cast order to (size_t)

File: oshmem/mca/spml/ucx/spml_ucx.c [hunk 5]
```diff
@@ -384,6 +384,34 @@ void mca_spml_ucx_rmkey_unpack(sshmem_mkey_t *mkey, uint32_t segno, int pe, int
     return;
 }
 
+void mca_spml_ucx_memuse_hook(void *addr, size_t length)
+{
+    int my_pe = oshmem_my_proc_id();
```
Comment (miked-mellanox) [inline oshmem/mca/spml/ucx/spml_ucx.c]: plz split into var and assignment

**Comment by alex-mikheev** [comment]
@yosefe @miked-mellanox can you rollback the merge of original comits to master ? I would rather fix it there and than cherry pick again

File: oshmem/mca/spml/ucx/spml_ucx.c [hunk 6]
```diff
@@ -384,6 +384,34 @@ void mca_spml_ucx_rmkey_unpack(sshmem_mkey_t *mkey, uint32_t segno, int pe, int
...
+                        UCP_MEM_ADVISE_PARAM_FIELD_LENGTH |
+                        UCP_MEM_ADVISE_PARAM_FIELD_ADVICE;
+
+    params.address = addr;
+    params.length  = length;
+    params.advice  = UCP_MADV_WILLNEED;
+
+    status = ucp_mem_advise(mca_spml_ucx.ucp_context, ucx_mkey->mem_h, &params);
+    if (UCS_OK != status) {
+        SPML_ERROR("ucp_mem_advise failed addr %p len %llu",
```
Comment (miked-mellanox) [inline oshmem/mca/spml/ucx/spml_ucx.c]: maybe warn? as it is not destructive?

Comment (alex-mikheev) [inline oshmem/mca/spml/ucx/spml_ucx.c]: There is no option for warning. It is either verbose or error.

**Comment by yosefe** [comment]
@alex-mikheev i think it would be easier to submit a fix commit both to master and to this PR

**Comment by miked-mellanox** [comment]
@alex-mikheev - please PR it into v2.x and v2.0

**Comment by hppritcha** [comment]
bot:lanl:retest

**Comment by jladd-mlnx** [comment]
@hppritcha Please merge this ASAP. The LANL failure looks unrelated to this commit.

**Comment by hppritcha** [comment]
bot:lanl:retest

**Pull request merged**